### PR TITLE
Add utilities for vector magnitude and normalisation

### LINKED ISSUE #190: Add norm function to the vector utils
**Is your feature request related to a problem? Please describe.**
We often need to get the magnitude of a vector.

**Describe the solution you'd like**
A `norm()` utility function that would function similar to existing `cart2pol` and `pol2cart` utilites.

This can facilitate #147

### PR DESCRIPTION
**Pull request by niksirbi**
## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
We often need to compute the magnitude (norm, length) of a vector, and would be nice to have a convenient xarray-native way of doing so. Currently we can only do that by either:
- use `cart2pol` to convert to poral coordinates, and take the `rho` value
- convert to numpy array and use `np.linalg.norm`

**What does this PR do?**
Adds two new functions in  the`utils/vector.py` module.
1. `magnitude`: this does the obvious thing. If an array in cartesian space is passed, we return the euclidean norm. If a user passes an array in polar coordinates, we just return the `rho`.
2. `normalize`: added for convenience. This function only works for arrays in cartesian coordinates, and returns the same array divided by its magnitude. For (0,0) values (exactly at the origin), the output is NaN (because of zero-division). If the input array is passed with polar coordinates, I raise an error. In principle, I could just make all `rho` values equal to 	1, but I don't see much value in that (feel free to push back if you have a better idea on what to do in this case).

## References

Closes #190.
Facilitates #147 , #239 , #228, #238 and others.

## How has this PR been tested?

Unit tests have been added for both functions.

## Is this a breaking change?

No.

## Does this PR require an update to the documentation?

I updated a few places in the examples where `np.linalg.norm` was used. Now they use `magnitude`.
The API index should automatically pick up the new functions.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [x] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

## PS
@sfmig I'm aware this was on your to-do list, but I had to do this anyway now, because we could benefit from this feature in https://github.com/neuroinformatics-unit/In2Research-2024.

## EDIT: 2024-08-14
Following @sfmig's review, the following changes were made:
- we renamed `magnitude()` to `compute_norm()` and updated the docstring accordingly
- we renamed `normalize()` to `convert_to_unit()` and updated the docstring accordingly
- we made `convert_to_unit()` also work in polar coordianates, where it makes all `rho` values 1 and preserves `phi` angles.

### CODE REVIEW

File: movement/utils/vector.py
Comment (lochhh) [inline movement/utils/vector.py:114]: These lines are repeated in `cart2pol()`, we should call `magnitude()` there directly instead.

**Comment by niksirbi** [comment]
@sfmig I'll give you a chance to comment on this before merging, because we've discussed this together. No rush

File: tests/test_unit/test_vector.py
Comment (sfmig) [inline tests/test_unit/test_vector.py:168]: maybe we could do this assertion more general: when the input vector is (0, 0), the corresponding norm should be NaN.

Right now we specifically check for the first frame.

File: tests/test_unit/test_vector.py
Comment (sfmig) [inline tests/test_unit/test_vector.py:174]: I like the comments before the assertions! But I find it easier to read this test if we add some spaces before these comments, to separate the assertions too. I feel this would help external contributors too.

**Review by sfmig — CHANGES_REQUESTED** [review]
Thanks for waiting for me @niksirbi and sorry for the late review! 

The implementation looks good, my suggestions are mostly around clarifying the docstrings. I also have two additional points.

I like that we stick to one term to refer to the norm of a vector. However, I suggest to use "norm" rather than "magnitude". The main reason for preferring "norm" is that that seems to be the preferred term for [numpy](https://numpy.org/doc/stable/reference/generated/numpy.linalg.vector_norm.html), [scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.norm.html), and [Matlab](https://uk.mathworks.com/help/matlab/ref/norm.html), which we have followed for the other vector utils (`cart2pol`, `pol2cart`). Many of our users will be familiar with these packages/language, and therefore it would make sense to keep the same nomenclature as much as possible.

But if we decide to go for "length" or "magnitude", I think we should accompany it with vector (vector length, vector magnitude) in the docstrings. I think this is currently not very clear because the input of the function is not a (2d, 3d) vector, but a multi-dimensional array, which suggests we are computing the magnitude of an array.

Re computing unit vectors for vectors in polar coordinates: I don't see a good reason for restricting the function to cartesian vectors only. As you say the implementation is easy, and the operation is conceptually the same for polar vectors. I think the function should do as expected (i.e., returning a vector with `rho=1` and same `phi`). 

Happy to discuss anything further!

File: movement/utils/vector.py
Comment (niksirbi) [reply movement/utils/vector.py:15]: I'll rename this function to `compute_norm` and rewrite the docstring as per your suggestions. Your version is indeed much clearer. I'll just move the last paragraph to "Notes":

```python
def compute_norm(data: xr.DataArray) -> xr.DataArray:
    """Compute the norm of the vectors along the spatial dimension.

    The norm of a vector is its magnitude, also called Euclidean norm, 2-norm
    or Euclidean length. Note that if the input data is expressed in polar
    coordinates, the magnitude of a vector is the same as its radial coordinate
    ``rho``. 

    Parameters
    ----------
    data : xarray.DataArray
        The input data array containing either ``space`` or ``space_pol``
        as a dimension.

    Returns
    -------
    xarray.DataArray
         A data array holding the norm of the input vectors.
         Note that this output array has no spatial dimension but preserves
         all other dimensions of the input data array (see Notes).

    Notes
    -----
    If the input data array is a ``position`` array, this function will compute
    the magnitude of the position vectors, for every individual and keypoint,
    at every timestep. If the input data array is a ``shape`` array of a 
    bounding boxes dataset, it will compute the magnitude of the shape
    vectors (i.e., the diagonal of the bounding box), 
    for every individual and at every timestep.
```

File: movement/utils/vector.py
Comment (niksirbi) [reply movement/utils/vector.py:27]: thanks for this, I was struggling with the phrasing.

File: movement/utils/vector.py
Comment (niksirbi) [reply movement/utils/vector.py:9]: You've convinced me for `compute_norm`

File: movement/utils/vector.py
Comment (niksirbi) [reply movement/utils/vector.py:50]: Quick quesiton @sfmig:

What do you think about `normalize_to_unit` or `convert_to_unit`? I feel like "compute" implies the production of a new quantity, while here I want to communicate that this functions doesn't produce something new but instead converts to a normalised format.

Also, do you think the inclusion of `vector` in the function name is necessary? I thought it might be superfluous because the whole module is named `vector.py`, but I might be overthinking this.

For context, I've decided to go with your `compute_norm` suggestion for the other function.

Comment (sfmig) [reply movement/utils/vector.py:50]: > What do you think about normalize_to_unit or convert_to_unit? 

I think `convert_to_unit` is clear and explicit, nice one! I don't feel strongly about "compute" but I see your point. 

> Also, do you think the inclusion of vector in the function name is necessary?

Yes I think you are right, it feels a bit redundant.

> I might be overthinking this.

I feel ya, it's kinda tricky 😅  (was there ~24h ago)

File: tests/test_unit/test_vector.py
Comment (niksirbi) [reply tests/test_unit/test_vector.py]: I've now changed this test quite a bit, taking into account all your comments:

```python
   @pytest.mark.parametrize(
        "ds, expected_exception",
        [
            ("cart_pol_dataset", does_not_raise()),
            ("cart_pol_dataset_with_nan", does_not_raise()),
            ("cart_pol_dataset_missing_cart_dim", pytest.raises(ValueError)),
        ],
    )
    def test_convert_to_unit(self, ds, expected_exception, request):
        """Test conversion to unit vectors (normalisation)."""
        ds = request.getfixturevalue(ds)
        with expected_exception:
            # normalise both the Cartesian and the polar data to unit vectors
            unit_cart = vector.convert_to_unit(ds.cart)
            unit_pol = vector.convert_to_unit(ds.pol)
            # they should yield the same result, just in different coordinates
            xr.testing.assert_allclose(unit_cart, vector.pol2cart(unit_pol))
            xr.testing.assert_allclose(unit_pol, vector.cart2pol(unit_cart))

            # since we established that polar vs Cartesian unit vectors are
            # equivalent, it's enough to do other assertions on either one

            # the normalised data should have the same dimensions as the input
            assert unit_cart.dims == ds.cart.dims

            # unit vector should be NaN if the input vector was null or NaN
            is_null_vec = (ds.cart == 0).all("space")  # null vec: x=0, y=0
            is_nan_vec = ds.cart.isnull().any("space")  # any NaN in x or y
            expected_nan_idxs = is_null_vec | is_nan_vec
            assert unit_cart.where(expected_nan_idxs).isnull().all()

            # For non-NaN unit vectors in polar coordinates, the rho values
            # should be 1 and the phi values should be the same as the input
            expected_unit_pol = ds.pol.copy()
            expected_unit_pol.loc[{"space_pol": "rho"}] = 1
            expected_unit_pol = expected_unit_pol.where(~expected_nan_idxs)
            xr.testing.assert_allclose(unit_pol, expected_unit_pol)
```

I think I now cover what you pointed out.

**Comment by niksirbi** [comment]
Thanks for the thorough review @sfmig. I believe I've addressed all your points by basically adopting all your suggestions (see edited part of the PR description). Re-requesting your review.

**Review by sfmig — APPROVED** [review]
looks great, thanks! ✨ 

I had a bit of a think about these tests, but I think they will be better addressed as a PR (draft [here](https://github.com/neuroinformatics-unit/movement/pull/274/files), still WIP) - will tag you when it's ready

**Pull request merged**
